### ISSUE TITLE
Validation error when using workloadRef

### ISSUE THREAD

**Issue by sidprak** [argoproj/argo-rollouts#1302]
## Summary 

This rollout works without any issues:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: test-service
spec:
  workloadRef:
    apiVersion: apps/v1
    kind: Deployment
    name: test-service
  strategy:
    canary:
      maxSurge: 10%
      maxUnavailable: 10%
      steps:
        - setWeight: 3
        - pause: {duration: 1m}
        - setWeight: 20
        - pause: {duration: 1m}
```

However, when I add analysis, like below, I get an incorrect error message. I would also expect this configuration to work. It just loops forever complaining that the rollout spec is not valid.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Rollout
metadata:
  name: test-service
spec:
  workloadRef:
    apiVersion: apps/v1
    kind: Deployment
    name: test-service
  strategy:
    canary:
      maxSurge: 10%
      maxUnavailable: 10%
      analysis:
        templates:
        - templateName: error-rate
        args:
        - name: service-name
          value: test-service
        - name: environment
          value: dev
      steps:
        - setWeight: 3
        - pause: {duration: 1m}
        - setWeight: 20
        - pause: {duration: 1m}
```

## Diagnostics

What version of Argo Rollouts are you running? **1.0.2**

These messages repeat forever and the rollout never completes.
```
time="2021-06-23T21:45:49Z" level=info msg="Started syncing rollout" generation=9 namespace=default resourceVersion=251502897 rollout=test-service
time="2021-06-23T21:45:49Z" level=info msg="Defaulting .spec.replica to 1" generation=9 namespace=default resourceVersion=251502897 rollout=test-service
time="2021-06-23T21:45:49Z" level=error msg="rollout syncHandler error: Rollout.argoproj.io \"test-service\" is invalid: [spec.selector: Invalid value: \"null\": spec.selector in body must be of type object: \"null\", spec.template.spec.containers: Invalid value: \"null\": spec.template.spec.containers in body must be of type array: \"null\"]" namespace=default rollout=test-service
time="2021-06-23T21:45:49Z" level=info msg="rollout syncHandler queue retries: 287 : key \"default/test-service\"" namespace=default rollout=test-service
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by huikang** [comment]
@sidprak , one workaround is to add `replicas: 1` in your rollout file.

```
  replicas: 1
  workloadRef:
    apiVersion: apps/v1
    kind: Deployment
    name: test-service
```

My guess is that the referred deployment has `replicas: 0`.

@jessesuen , I can look into the issue; please assign it to me. Thanks.

**Comment by huikang** [comment]
@sidprak , could you paste the deployment manifest the rollout refers? Thanks.

**Comment by sidprak** [comment]
@huikang thanks for the quick reply! I can confirm that setting `replicas: 1` in the rollout does work but I run into another issue with that. We use HPA on this rollout and we use ArgoCD for deployment. With this setup, the HPA modifies the # of replicas to something like 10, and ArgoCD constantly tries to set it back to 1. That's the main reason I removed the `replicas` setting from the rollout. Let me know if I should open a separate issue for this.

This is the deployment manifest:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    argocd.argoproj.io/sync-wave: '2'
  labels:
    app: test-service
    appInst: test-service
  name: test-service
spec:
  progressDeadlineSeconds: 600
  replicas: 0
  selector:
    matchLabels:
      appInst: test-service
  template:
    metadata:
      labels:
        app: test-service
        appInst: test-service
    spec:
      containers:
      - image: test-service:dev
        lifecycle:
          preStop:
            exec:
              command:
              - /bin/sleep
              - '25'
        name: test-service
        ports:
        - containerPort: 5000
        resources:
          limits:
            cpu: 200m
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 100Mi
        volumeMounts:
        - mountPath: /app/config
          name: config
      securityContext:
        fsGroup: 1000
      serviceAccountName: test-service
      volumes:
      - configMap:
          name: test-service-config
        name: config

```

**Comment by huikang** [comment]
@sidprak , thanks for your confirmation. I am working on a PR to resolve the issue of not defaulting the replica if the field does not appear in the rollout manifest. Hopefully it will resolve the issue you just described; otherwise we can open up a separate issue.

Will let you know when the PR is ready for testing.

**Comment by huikang** [comment]
@sidprak , could you try the fix in the PR: #1304 ? I tested locally and it worked. Thanks.

**Comment by sidprak** [comment]
@huikang your fix works for me! Thanks for fixing it so quickly.
